momaland/envs/gem_mining/gem_mining.py

```diff
@@ -0,0 +1,274 @@
+        self.observation_spaces = dict(
+            zip(
+                self.agents,
+                [
+                    Box(
+                        low=0,
+                        high=self.num_agents,
+                        shape=(1,),
+                        dtype=np.float32,
+                    )
+                ]
+                * num_agents,
+            )
+        )
+        self.reward_spaces = dict(zip(self.agents, [Box(low=0, high=self.num_mines, shape=(num_objectives,))] * num_agents))
+
+        # set truncations to false (no agents are ever lost)
+        self.truncations = {agent: False for agent in self.agents}
+
+    # this cache ensures that same space object is returned for the same agent
+    # allows action space seeding to work as expected
+    @functools.lru_cache(maxsize=None)
+    @override
+    def observation_space(self, agent):
+        return self.observation_spaces[agent]
+
+    @functools.lru_cache(maxsize=None)
+    @override
+    def action_space(self, agent):
+        return self.action_spaces[agent]
+
+    @override
+    def reward_space(self, agent):
+        return self.reward_spaces[agent]
+
+    @override
+    def render(self):
+        if self.render_mode is None:
+            warn("You are calling render method without specifying any render mode.")
+            return
+
+    @override
+    def close(self):
+        """Close should release any graphical displays, subprocesses, network connections or any other environment data which should not be kept around after the user is no longer using the environment."""
+        pass
+
+    @override
+    def reset(self, seed=None, options=None):
+        """Reset needs to initialize the `agents` attribute and must set up the environment so that render(), and step() can be called without issues.
+
+        Returns the observations for each agent
+        """
+        if seed is not None:
+            print(f"seeding with seed {seed}")
+            self.np_random.seed(seed)
+            self.random.seed(seed)
+        self.agents = self.possible_agents[:]
+        self.terminations = {agent: False for agent in self.agents}
+        self.truncations = {agent: False for agent in self.agents}
+        observations = {agent: np.array([0], dtype=np.float32) for i, agent in enumerate(self.agents)}
+        self.episode_num = 0
+
+        infos = {agent: {} for agent in self.agents}
+        return observations, infos
+
+    def step(self, actions):
+        """Steps in the environment.
+
+        Args:
+            actions: a dict of actions, keyed by agent names
+
+        Returns: a tuple containing the following items in order:
+        - observations
+        - rewards
+        - terminations
+        - truncations
+        - infos
+        dicts where each dict looks like {agent_1: item_1, agent_2: item_2}
+        """
+        # If a user passes in actions with no agents, then just return empty observations, etc.
+        if not actions:
```

Comment:
I have not seen this before. I think this also only gets triggered if actions is `False` which I do not see a use-case for?

Reply:
No clue either